test/tests/functional/pbs_offline_vnodes.py

```diff
@@ -0,0 +1,204 @@
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class TestOfflineVnodeOnHookFailure(TestFunctional):
+    """
+    Tests if vnodes are marked offline if a hook fails
+    """
+
+    def setUp(self):
+        if len(self.moms.values()) != 2:
+            self.skipTest("Provide 2 moms while invoking test")
+        if not self.du.get_platform().startswith('cray'):
+            self.is_cray = False
+        else:
+            self.is_cray = True
+
+        # The moms provided to the test may have unwanted vnodedef files.
+        if self.moms.values()[0].has_vnode_defs():
+            self.moms.values()[0].delete_vnode_defs()
+        if self.moms.values()[1].has_vnode_defs():
+            self.moms.values()[1].delete_vnode_defs()
+        TestFunctional.setUp(self)
+
+    def create_mom_hook(self):
+        name = "h1"
+        body = ("import pbs\n"
+                "import time\n"
+                "time.sleep(60)\n"
+                "pbs.event.accept()")
+        attr = {'event': 'execjob_begin', 'fail_action': 'offline_vnodes',
+                'alarm': '3', 'enabled': 'true'}
+        self.server.create_import_hook(name, attr, body)
+
+    def create_multi_vnodes(self, num_moms, opt):
+        if self.is_cray is True:
+            return
+        vn_attrs = {ATTR_rescavail + '.ncpus': 1,
+                    ATTR_rescavail + '.mem': '1024mb'}
+        for i in range(num_moms):
+            self.server.create_vnodes('vnode', vn_attrs, 3, self.moms.values()[
+                                      i], usenatvnode=True, delall=opt,
+                                      expect=False)
+
+    def test_single_mom_hook_failure_affects_vnode(self):
+        """
+        Run an execjob_begin hook that sleep for sometime,
+        at the same time set an alarm value so less that
+        the hook alarms out and server executes the fail_action
+        After this check if vnodes are marked offline.
+        In case of a single mom reporting vnodes, it should mark
+        all the vnodes and mom as offline.
+        """
+        self.create_multi_vnodes(1, True)

```

Comment:
In the case of Cray, you start out with two login nodes. You might want to disable one of them so that your test will truly have one mom.

Reply:
This single mom test might be good to run on the Altix/UV. But the setup requires two moms.